src/postgres-util/src/privileges.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use anyhow::anyhow;
+use postgres_array::{Array, Dimension};
+
+use crate::{Config, PostgresError};
+
+/// Ensure that the user specified in `config` has `SELECT` privileges for the identified `tables`.
+///
+/// `tables`'s elements should be of the structure `[<schema name>, <table name>]`.
+///
+/// # Panics
+/// If `config` does not specify a user.
+pub async fn check_select_privileges(
+    config: &Config,
+    tables: Vec<[&str; 2]>,
+) -> Result<(), PostgresError> {
+    let client = config.connect("check_privileges").await?;
+
+    let tables_len = tables.len();
+
+    let tables = Array::from_parts(
+        tables.into_iter().map(|i| i.to_vec()).flatten().collect(),
+        vec![
+            Dimension {
+                len: i32::try_from(tables_len).expect("fewer than i32::MAX tables"),
+                lower_bound: 0,
+            },
+            Dimension {
+                len: 2,
+                lower_bound: 0,
+            },
+        ],
+    );
+
+    let invalid_read_privileges = client
+        .query(
+            "
+            SELECT
+                table_catalog || '.' || table_schema || '.' || table_name AS t
+            FROM information_schema.table_privileges
+            WHERE
+                ARRAY[table_schema::text, table_name]
+                <@ $1
+                    AND
+                table_catalog = $2 AND
+                grantee = $3
+            GROUP BY table_catalog, table_schema, table_name
+            HAVING NOT 'SELECT' =ANY (array_agg(privilege_type::text))
+            ORDER BY t;",
+            &[
+                &tables,
+                &config.get_dbname(),
+                &config.get_user().expect("connection specifies user"),
+            ],
+        )
+        .await?
+        .into_iter()
+        .map(|row| row.get("t"))
+        .collect::<Vec<String>>();
+
+    if invalid_read_privileges.is_empty() {
+        Ok(())
```

Comment:
Can this cause a race condition where the user loses these privileges after checking but before the source is created?

Reply:
Yes, but that is always going to be the case unless we jump through a bunch of hoops to not parameterize the query so we can run it inside the transaction that snapshots.